cmd/xl-v1-object.go

```diff
@@ -42,13 +42,105 @@ var objectOpIgnoredErrs = []error{
+	// Do we have read quorum?
+	if !isDiskQuorum(errs, xl.readQuorum) {
+		return ObjectInfo{}, traceError(InsufficientReadQuorum{}, errs...)
+	}
+
+	if reducedErr := reduceReadQuorumErrs(errs, objectOpIgnoredErrs, xl.readQuorum); reducedErr != nil {
+		return ObjectInfo{}, toObjectErr(reducedErr, srcBucket, srcObject)
+	}
+
+	// List all online disks.
+	onlineDisks, modTime := listOnlineDisks(xl.storageDisks, metaArr, errs)
+
+	// Pick latest valid metadata.
+	xlMeta, err := pickValidXLMeta(metaArr, modTime)
+	if err != nil {
+		return ObjectInfo{}, toObjectErr(err, srcBucket, srcObject)
+	}
+
+	// Reorder online disks based on erasure distribution order.
+	onlineDisks = getOrderedDisks(xlMeta.Erasure.Distribution, onlineDisks)
+
+	// Length of the file to read.
+	length := xlMeta.Stat.Size
+
+	// Check if this request is only metadata update.
+	cpMetadataOnly := strings.EqualFold(pathJoin(srcBucket, srcObject), pathJoin(dstBucket, dstObject))
+	if cpMetadataOnly {
+		xlMeta.Meta = metadata
+		partsMetadata := make([]xlMetaV1, len(xl.storageDisks))
+		// Update `xl.json` content on each disks.
+		for index := range partsMetadata {
+			partsMetadata[index] = xlMeta
+		}
+
+		tempObj := mustGetUUID()
+
+		// Write unique `xl.json` for each disk.
+		if err = writeUniqueXLMetadata(onlineDisks, minioMetaTmpBucket, tempObj, partsMetadata, xl.writeQuorum); err != nil {
+			return ObjectInfo{}, toObjectErr(err, srcBucket, srcObject)
+		}
+		// Rename atomically `xl.json` from tmp location to destination for each disk.
+		if err = renameXLMetadata(onlineDisks, minioMetaTmpBucket, tempObj, srcBucket, srcObject, xl.writeQuorum); err != nil {
+			return ObjectInfo{}, toObjectErr(err, srcBucket, srcObject)
+		}
+
+		objInfo := ObjectInfo{
+			IsDir:           false,
+			Bucket:          srcBucket,
+			Name:            srcObject,
+			Size:            xlMeta.Stat.Size,
+			ModTime:         xlMeta.Stat.ModTime,
+			MD5Sum:          xlMeta.Meta["md5Sum"],
+			ContentType:     xlMeta.Meta["content-type"],
+			ContentEncoding: xlMeta.Meta["content-encoding"],
+		}
+		// md5Sum has already been extracted into objInfo.MD5Sum.  We
+		// need to remove it from xlMetaMap to avoid it from appearing as
+		// part of response headers. e.g, X-Minio-* or X-Amz-*.
+		delete(xlMeta.Meta, "md5Sum")
+		objInfo.UserDefined = xlMeta.Meta
+		return objInfo, nil
+	}
+
+	// Initialize pipe.
+	pipeReader, pipeWriter := io.Pipe()
+
+	go func() {
+		startOffset := int64(0) // Read the whole file.
+		if gerr := xl.GetObject(srcBucket, srcObject, startOffset, length, pipeWriter); gerr != nil {
+			errorIf(gerr, "Unable to read %s of the object `%s/%s`.", srcBucket, srcObject)
+			pipeWriter.CloseWithError(toObjectErr(gerr, srcBucket, srcObject))
+			return
+		}
+		pipeWriter.Close() // Close writer explicitly signalling we wrote all data.
+	}()
+
+	objInfo, err := xl.PutObject(dstBucket, dstObject, length, pipeReader, xlMeta.Meta, "")
```

Comment:
metadata needs to be considered instead of xlMeta.Meta.
the right metadata will be passed by the S3 layer depending on what the `x-amz-metadata-directive` is.